### DISCUSSION TITLE
Comportamiento inusual e inexplicable de fprintf

### DISCUSSION THREAD

**Discussion by maxcareyf** [IIC2133-PUC/2024-2#67] [Tarea 0] [answered]
Estoy teniendo problemas con el funcionamiento de mi tarea en relación con la escritura al archivo de salida, y necesito la ayuda de un mago de C. Después de terminar mi programa, el primer par de pruebas funcionaron perfectamente, pero repentinamente falló con una prueba fácil. Noté que el fallo fue que paró de escribir a output_file repentinamente, por lo que pensé que debería tener un error en el conteo de comandos, por lo cual terminó antes. Sin embargo, al probar otros tests, el programa nunca paró de ejecutarse, por lo cual pensé que definitivamente era un mal manejo de las iteraciones. Incluso, en uno de los tests, recibí un error de segmentation fault, lo cual también me hizo pensar en mal manejo de memoria. Para resolver mis dudas, agregué un printf (NO fprintf) al inicio de cada iteración que indica cuantos comandos quedaban por ejecutar, y lo sorprendente de esto es que ahora el programa siempre termina sin errores y en la cantidad de comandos correcta. Valgrind indica que todos los tests tienen 0 errores y toda la memoria en todos los tests es liberada por completo, lo cual me dice que el programa se está ejecutando en su totalidad, y ahora para variar me pasa que al archivo de output donde escribo la salida siempre se le escribe exactamente una línea de salida (correcta :) ). Intenté agregar fflush al final de cada fprintf, pero no solucionó nada. De intuición se me ocurre mal manejo de buffer de fprintf (si es que existe) pero realmente no sé qué pasa con mi tarea. Le ruego ayuda a algún mago de C. Muchas gracias.

**Comment by tomastrivino** [comment, marked as answer]
No soy un dios de C ni un mago de C, pero me gustaría añadir la consideración de que se han visto varios problemas por el asunto del `char color` en el código base (en vez de char *color o char color[NRO DIGITOS]. Si esta parte ya la tienes arreglada, avísanos y asegúrate de tener tu tarea en el repositorio para poder probarla

**Reply by maxcareyf** [reply]
Hola! En mi struct de Puffle, color es un puntero a char al cual siempre se le asignan 20 chars en el heap, mientras que para BUY-BUFFLE, ENTER y STEAL-PUFFLE el input es char[20]. Siempre ocupo strcpy para copiar el color a la dirección de color, y según valgrind no hay errores ni leaks de memoria. Mi código está en mi repo, no modularizado pero si segmentado.

**Reply by tomastrivino** [reply]
Lo probé sin mayores problemas. El único comentario que tengo es que el print de `commands pending` lo pusiste justo abajo de la línea del for. Debe ir dentro de los corchetes si quieres que esté en el for.
Saludos

**Reply by tomastrivino** [reply]
De hecho, sin ese print de debuggeo, me dieron inputs correctos para los primeros dos test.
